docs/spec/fastsync/fastsync.md

```diff
@@ -0,0 +1,1163 @@
+has not provided sufficiently many data (check of *peerRate[p]*), then
+*p* is removed from *peerIDs*.
+  
+#### **[FS-V2-TIMEOUT]**:
+
+*Fastsync V2* starts a timeout whenever a block is
+executed (that is, when the height is incremented). If the timeout expires
+before the next block is executed, *Fastsync* terminates.
+We say that if *peerIDs* is empty upon termination, then *Fastsync* terminates
+with failure, otherwise it terminates successfully.
+
+
+
+
+### Details
+
+<!--
+> Function signatures followed by pseudocode (optional) and a list of features (required):
+> - Implementation remarks (optional)
+>   - e.g. (local/remote) function called in the body of this function
+> - Expected precondition
+> - Expected postcondition
+> - Error condition
+---->
+
+
+
+```go
+func QueryStatus()
+```
+- Expected precondition
+    - peerIDs initialized and non-empty
+- Expected postcondition
+    - call asynchronously `Status(n)` at each peer *n* in *peerIDs*.
+- Error condition
+    - fails if precondition is violated
+----
+
+```go
+func OnStatusResponse(addr Address, ht int64)
+```
+- Comment
+    - *ht* is a height
+	- peers can provide the status without being called
+- Expected precondition
+    - *peerHeights(addr) <= ht*
+- Expected postcondition
+    - *peerHeights(addr) = ht*
+	- *TargetHeight* is updated
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs* (that is,
+      *addr* is removed from *peerIDs*)
+- Timeout condition
+    - if `OnStatusResponse(addr, ht)` was not invoked within *2 Delta* after
+	`Status(addr)` was called:  *addr* not in *peerIDs*
+----
+
+
+```go
+func CreateRequest
+```
+- Expected precondition
+    - *height < TargetHeight*
+	- *peerIDs* nonempty
+- Expected postcondition
+    - Function `Block` is called remotely at a peer *addr* in peerIDs 
+	  for a missing height *h*  
+	  *Remark:* different implementations may have different
+      strategies to balance the load over the peers
+    - *pendingblocks(h) = addr*
+----
+
+
+```go
+func OnBlockResponse(addr Address, b Block)
+```
+- Comment
+    - if after adding block *b*, blocks of heights *height* and
+      *height + 1* are in *blockstore*, then `Execute` is called
+- Expected precondition
+    - *pendingblocks(b.Height) = addr*
+	- *b* satisfies basic soundness  
+- Expected postcondition
+    - if function `Execute` has been executed without error or was not
+      executed:
+        - *receivedBlocks(b.Height) = addr*
+        - *blockstore(b.Height) = b*
+        - *peerTimeStamp[addr]* is set to a time between invocation and
+          return of the function.
+        - *peerRate[addr]* is updated according to size of received
+          block
+- Error condition
+    - if precondition is violated: *addr* not in *peerIDs*; reset
+	*pendingblocks(b.Height)* to nil;
+- Timeout condition
+    - if `OnBlockResponse(addr, b)` was not invoked within *2 Delta* after
```

Comment:
In v2, there is no such check. Removal happens periodically but only for the current height this kind of check is implemented. For other peers, it looks at peer rate and difference between current time and last block reception.